### PR TITLE
JavaScript and Python unit tests

### PR DESCRIPTION
**Pull request by jdtzmn**
### Added

- JavaScript testing using karma
- Registration unit tests
- API unit tests
- Scaffold for integration tests (not yet in use)

### CODE REVIEW

File: flocx_ui/test/integration/horizon.conf [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[plugin]
+is_plugin=True
+plugin_page_path=flocx_ui.test.integration.pages
+plugin_page_structure='{"Project": {"Other": {"_": ["Flocx"]}}}'
```
Comment (gagansk) [inline flocx_ui/test/integration/horizon.conf]: Add a new line. It is a nice practice to add '\n' which will help with diffs and EOF later.

File: .travis.yml [hunk 2]
```diff
@@ -1,12 +1,22 @@
...
-      script:
-        - yarn lint
-    - language: python
-      dist: xenial
-      python: 3.7
-      install:
-        - pip install -r requirements.txt -e .
+    - name: JavaScript
+      script: yarn test
+    - name: Python
```
Comment (gagansk) [inline .travis.yml:15]: Are you using Python 3.7?

File: flocx_ui/test/integration/pages/admin/other/flocxpage.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from openstack_dashboard.test.integration_tests.pages import basepage
+
+
+class FlocxPage(basepage.BaseNavigationPage):
+    def __init__(self, driver, conf):
+        super(FlocxPage, self).__init__(driver, conf)
+        self._page_title = "Flocx"
```
Comment (gagansk) [inline flocx_ui/test/integration/pages/admin/other/flocxpage.py]: Add a new line. It is a nice practice to add '\n' which will help with diffs and EOF later.

File: flocx_ui/test/integration/test_basic.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from openstack_dashboard.test.integration_tests import helpers
+
+
```
Comment (gagansk) [inline flocx_ui/test/integration/test_basic.py:3]: Please define the purpose of the test

File: flocx_ui/test/tests/test_flocx_rest_api.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from unittest import mock
+
+from openstack_dashboard.test import helpers as test
+
+from flocx_ui.api.flocx_rest_api import Offer as OfferAPI
+from flocx_ui.test.tests.helpers import get_test_data, RequestFactory
+
+class RestApiTests(test.TestCase):
```
Comment (gagansk) [inline flocx_ui/test/tests/test_flocx_rest_api.py:8]: Define the definition and scope of the test

File: flocx_ui/test/tests/test_flocx_rest_api.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    def test_get_offers(self, mock_offer_list):
+        testData = get_test_data('offer_list_response')
+        mock_offer_list.return_value = testData
+
+        rf = RequestFactory()
+        request = rf.get('/api/flocx/offer/')
+
+        offerAPI = OfferAPI()
+        response = offerAPI.get(request)
+        self.assertEqual(response.json, testData)
```
Comment (gagansk) [inline flocx_ui/test/tests/test_flocx_rest_api.py:28]: Add negative tests as well

**Review by gagansk — CHANGES_REQUESTED** [review]
Good job! A couple of comments:
- Please try to add negative test whenever it is possible. 
- Always define the test and the expected output. This will help to know which part of the component might be broken as soon as a test fails since assert will not provide the answer immediately.

File: .travis.yml (hunk 2 above)
Comment (jdtzmn) [reply .travis.yml:15]: Yes. Should I be using a different version?

File: flocx_ui/test/tests/test_flocx_rest_api.py (hunk 6 above)
Comment (jdtzmn) [reply flocx_ui/test/tests/test_flocx_rest_api.py:28]: Handling of those sorts of errors is done automatically using the `@rest_utils.ajax()` decorator. You can see where that is implemented in the [`flocx_rest_api.py` file](https://github.com/jdtzmn/flocx-ui/blob/add-tests/flocx_ui/api/flocx_rest_api.py) or view the actual decorator source code [here](https://github.com/openstack/horizon/blob/master/openstack_dashboard/api/rest/utils.py#L76).

I do write negative tests where applicable (see the `test_create_invalid_offer` test in [this file](https://github.com/CCI-MOC/flocx-ui/pull/7/files#diff-060a18ea8795e0578bf89fa9153088e5R40) in PR #7), but I’m not sure that it makes sense here.

Let me know if this is not what you meant.

File: .travis.yml (hunk 2 above)
Comment (gagansk) [reply .travis.yml:15]: No. That's good. I just wanted to confirm.

File: flocx_ui/test/tests/test_flocx_rest_api.py (hunk 6 above)
Comment (gagansk) [reply flocx_ui/test/tests/test_flocx_rest_api.py:28]: Yes. That is what I was thinking

**Review by gagansk — APPROVED** [review]

**Comment by gagansk** [comment]
lgtm

**Pull request merged**